File: core/src/net/artux/pda/map/engine/systems/player/PlayerMovingSystem.java
package net.artux.pda.map.engine.systems.player;

import com.badlogic.ashley.core.ComponentMapper;
import com.badlogic.ashley.core.Entity;
import com.badlogic.ashley.core.Family;
import com.badlogic.gdx.assets.AssetManager;
import com.badlogic.gdx.audio.Sound;
import com.badlogic.gdx.math.Vector2;

import net.artux.pda.map.DataRepository;
import net.artux.pda.map.di.scope.PerGameMap;
import net.artux.pda.map.engine.components.HealthComponent;
import net.artux.pda.map.engine.components.PassivityComponent;
import net.artux.pda.map.engine.components.Position;
import net.artux.pda.map.engine.components.VelocityComponent;
import net.artux.pda.map.engine.data.GlobalData;
import net.artux.pda.map.engine.pathfinding.TiledNode;
import net.artux.pda.map.engine.systems.BaseSystem;
import net.artux.pda.map.engine.systems.MapOrientationSystem;
import net.artux.pda.model.quest.story.StoryDataModel;

import org.apache.commons.lang3.tuple.ImmutablePair;

import java.util.HashMap;
import java.util.Random;

import javax.inject.Inject;

@PerGameMap
public class PlayerMovingSystem extends BaseSystem {

    private final float MOVEMENT = 20f;
    private final float RUN_MOVEMENT = 30f;
    private final float PLAYER_MULTIPLICATION = 6f;

    private ComponentMapper<Position> pm = ComponentMapper.getFor(Position.class);
    private ComponentMapper<VelocityComponent> vm = ComponentMapper.getFor(VelocityComponent.class);
    private ComponentMapper<HealthComponent> hm = ComponentMapper.getFor(HealthComponent.class);

    private MapOrientationSystem mapOrientationSystem;

    public static boolean playerWalls = true;
    public static boolean speedup = false;
    public static boolean alwaysRun = false;

    private float weightCoefficient = 0f;

    private final HashMap<Integer, ImmutablePair<Sound, Sound>> stepSounds;
    private boolean left = false;
    private final float stepVolume = 0.15f;
    private final float oneSoundDistance = 5.7f;
    private float stepsDistance = 0;
    private final Random random;

    @Inject
    public PlayerMovingSystem(AssetManager assetManager, MapOrientationSystem mapOrientationSystem, DataRepository dataRepository) {
        super(Family.all(VelocityComponent.class, Position.class).exclude(PassivityComponent.class).get());
        this.mapOrientationSystem = mapOrientationSystem;
        dataRepository.addPropertyChangeListener(propertyChangeEvent -> {
            StoryDataModel storyDataModel = (StoryDataModel) propertyChangeEvent.getNewValue();
            weightCoefficient = 1.5f - storyDataModel.getTotalWeight() / 60;
        });

        weightCoefficient = 1.5f - dataRepository.getStoryDataModel().getTotalWeight() / 60;
        if (weightCoefficient < 0.1f)
            weightCoefficient = 0.1f;
        stepSounds = new HashMap<>();
        random = new Random();

        String prefix = "audio/sounds/steps/";

        stepSounds.put(TiledNode.TILE_EMPTY,
                ImmutablePair.of(assetManager.get(prefix + "empty1.ogg"), assetManager.get(prefix + "empty2.ogg")));

        stepSounds.put(TiledNode.TILE_ROAD,
                ImmutablePair.of(assetManager.get(prefix + "road1.ogg"), assetManager.get(prefix + "road2.ogg")));

        stepSounds.put(TiledNode.TILE_GRASS,
                ImmutablePair.of(assetManager.get(prefix + "grass1.ogg"), assetManager.get(prefix + "grass2.ogg")));

        stepSounds.put(TiledNode.TILE_SWAMP,
                ImmutablePair.of(assetManager.get(prefix + "swamp1.ogg"), assetManager.get(prefix + "swamp2.ogg")));
    }

    @Override
    public void update(float deltaTime) {
        super.update(deltaTime);
        Entity entity = getPlayer();
        Position position = pm.get(entity);
        VelocityComponent velocityComponent = vm.get(entity);

        velocityComponent.setVelocity(velocityComponent.cpy());
        if (alwaysRun)
            velocityComponent.setRunning(true);

        Vector2 stepVector;
        Vector2 currentVelocity = velocityComponent.cpy().scl(weightCoefficient);
        if (speedup)
            currentVelocity.scl(PLAYER_MULTIPLICATION);

        HealthComponent healthComponent = hm.get(entity);

        float staminaDifference = 0;
        if (velocityComponent.isRunning() && healthComponent.stamina > 0) {
            stepVector = currentVelocity.scl(deltaTime).scl(RUN_MOVEMENT);
            if (!alwaysRun && entity == getPlayer())
                staminaDifference = -0.1f;
        } else {
            if (healthComponent.stamina < 100)
                staminaDifference = 0.06f;
            stepVector = currentVelocity.scl(deltaTime).scl(MOVEMENT);
        }

        healthComponent.stamina += staminaDifference;

        if (!stepVector.isZero()) {
            float newX = position.getX() + stepVector.x;
            float newY = position.getY() + stepVector.y;
            if (playerWalls) {
                if (insideMap(newX, position.getY()))
                    position.getPosition().x += stepVector.x * mapOrientationSystem.getMapBorder().getK(newX, position.getY());
                if (insideMap(position.getX(), newY))
                    position.getPosition().y += stepVector.y * mapOrientationSystem.getMapBorder().getK(position.getX(), newY);
            } else {
                if (insideMap(newX, position.getY()))
                    position.getPosition().x = newX;
                if (insideMap(position.getX(), newY))
                    position.getPosition().y = newY;
            }

            stepsDistance += stepVector.len();
            if (stepsDistance >= oneSoundDistance) {
                stepsDistance = 0;
                int type = mapOrientationSystem.getMapBorder().getTileType(position.x, position.y);
                if (!stepSounds.containsKey(type) || random.nextInt(4) == 0)
                    type = TiledNode.TILE_EMPTY;

                if (left)
                    stepSounds.get(type).left.play(stepVolume * currentVelocity.len());
                else
                    stepSounds.get(type).right.play(stepVolume * currentVelocity.len());
                left = !left;
            }
        }
    }

    @Override
    protected void processEntity(Entity entity, float deltaTime) {

    }

    public boolean insideMap(float x, float y) {
        return (x <= GlobalData.mapWidth && x >= 0) && (y <= GlobalData.mapHeight && y >= 0);
    }

}

File: core/src/net/artux/pda/map/engine/ecs/systems/player/PlayerMovingSystem.kt
package net.artux.pda.map.engine.ecs.systems.player

import com.badlogic.ashley.core.ComponentMapper
import com.badlogic.ashley.core.Entity
import com.badlogic.ashley.core.Family
import com.badlogic.gdx.assets.AssetManager
import com.badlogic.gdx.audio.Sound
import com.badlogic.gdx.math.Vector2
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import net.artux.engine.pathfinding.TiledNode
import net.artux.pda.map.DataRepository
import net.artux.pda.map.di.scope.PerGameMap
import net.artux.pda.map.engine.ecs.components.BodyComponent
import net.artux.pda.map.engine.ecs.components.HealthComponent
import net.artux.pda.map.engine.ecs.components.PassivityComponent
import net.artux.pda.map.engine.ecs.components.VelocityComponent
import net.artux.pda.map.engine.ecs.systems.BaseSystem
import net.artux.pda.map.engine.ecs.systems.MapOrientationSystem
import org.apache.commons.lang3.tuple.ImmutablePair
import java.util.*
import javax.inject.Inject

@PerGameMap
class PlayerMovingSystem @Inject constructor(
    val assetManager: AssetManager,
    val mapOrientationSystem: MapOrientationSystem,
    val dataRepository: DataRepository
) : BaseSystem(
    Family
        .all(VelocityComponent::class.java, BodyComponent::class.java)
        .exclude(PassivityComponent::class.java).get()
) {

    private val pm = ComponentMapper.getFor(BodyComponent::class.java)
    private val vm = ComponentMapper.getFor(VelocityComponent::class.java)
    private val hm = ComponentMapper.getFor(HealthComponent::class.java)

    private val MOVEMENT = 20f
    private val RUN_MOVEMENT = 30f
    private val PLAYER_MULTIPLICATION = 6f

    private var weightCoefficient = 0f
    private var stepSounds: HashMap<Int, ImmutablePair<Sound, Sound>>
    private var left = false
    private val stepVolume = 0.15f
    private val oneSoundDistance = 5.7f
    private var stepsDistance = 0f
    private var random: Random

    override fun update(deltaTime: Float) {
        super.update(deltaTime)

        val entity = player
        val position = pm[entity]
        val velocityComponent = vm[entity]
        velocityComponent.velocity = velocityComponent.cpy()

        if (alwaysRun) velocityComponent.isRunning = true
        val stepVector: Vector2
        val currentVelocity = velocityComponent.cpy().scl(weightCoefficient)
        if (speedup) currentVelocity.scl(PLAYER_MULTIPLICATION)
        val healthComponent = hm[entity]
        var staminaDifference = 0f
        if (velocityComponent.isRunning && healthComponent.stamina > 0) {
            stepVector = currentVelocity.scl(deltaTime).scl(RUN_MOVEMENT)
            if (!alwaysRun && entity === player) staminaDifference = -0.1f
        } else {
            if (healthComponent.stamina < 100) staminaDifference = 0.06f
            stepVector = currentVelocity.scl(deltaTime).scl(MOVEMENT)
        }
        healthComponent.stamina += staminaDifference
        if (!stepVector.isZero) {
            position.getBody().applyLinearImpulse(
                stepVector.x * MOVEMENT,
                stepVector.y * MOVEMENT,
                position.x,
                position.y,
                true
            )
        }
        stepsDistance += stepVector.len()
        if (stepsDistance >= oneSoundDistance) {
            stepsDistance = 0f
            var type = mapOrientationSystem.mapBorder.getTileType(position.x, position.y)
            if (!stepSounds.containsKey(type) || random.nextInt(4) == 0) type =
                TiledNode.TILE_EMPTY
            if (left) stepSounds[type]!!.left.play(stepVolume * currentVelocity.len()) else stepSounds[type]!!.right.play(
                stepVolume * currentVelocity.len()
            )
            left = !left
        }

    }

    override fun processEntity(entity: Entity, deltaTime: Float) {}

    companion object {
        var playerWalls = true
        var speedup = false
        var alwaysRun = false
    }

    init {
        CoroutineScope(Dispatchers.Main).launch {
            dataRepository.storyDataModelFlow.collect {
                val storyDataModel = it
                weightCoefficient = 1.5f - storyDataModel.totalWeight / 60
            }
        }

        weightCoefficient = 1.5f - dataRepository.storyDataModel.totalWeight / 60
        if (weightCoefficient < 0.1f) weightCoefficient = 0.1f
        stepSounds = HashMap()
        random = Random()
        val prefix = "audio/sounds/steps/"
        stepSounds[TiledNode.TILE_EMPTY] =
            ImmutablePair.of(
                assetManager.get(prefix + "empty1.ogg"),
                assetManager.get(prefix + "empty2.ogg")
            )
        stepSounds[TiledNode.TILE_ROAD] = ImmutablePair.of(
            assetManager.get(prefix + "road1.ogg"),
            assetManager.get(prefix + "road2.ogg")
        )
        stepSounds[TiledNode.TILE_GRASS] = ImmutablePair.of(
            assetManager.get(prefix + "grass1.ogg"),
            assetManager.get(prefix + "grass2.ogg")
        )
        stepSounds[TiledNode.TILE_SWAMP] =
            ImmutablePair.of(
                assetManager.get(prefix + "swamp1.ogg"),
                assetManager.get(prefix + "swamp2.ogg")
            )
    }
}